scriptoocloud/src/main/java/edu/asu/diging/scriptoocloud/core/service/impl/DatasetService.java

```diff
@@ -0,0 +1,193 @@
+        dataset.setName(name);
+        dataset.setUser(user);
+        Dataset savedDataset = datasetRepository.save(dataset);
+        Long savedDatasetId = savedDataset.getId();
+
+        // create directories on the file system (using Dataset id)
+        try {
+            fileSystemService.addDatasetDirectories(String.valueOf(savedDatasetId),
+                    user.getUsername());
+        } catch (DatasetStorageException e) {
+            throw new DatasetStorageException(
+                    "An IOException prevented the Dataset from being created", e);
+        }
+    }
+
+    /**
+     * Edits the dataset name in the database (Dataset path uses id).
+     *
+     * @param id       The id of the dataset to edit.
+     * @param newName  The new name of the dataset being edited.
+     */
+    @Override
+    public void editDataset(Long id, String newName) {
+
+        // filesystem path is: username/datasetId so no change is needed to the filesystem
+        Optional<Dataset> dataset = datasetRepository.findById(id);
+        if (dataset.isPresent()) {
+            dataset.get().setName(newName);
+            datasetRepository.save(dataset.get());
+        }
+    }
+
+    /**
+     * Deletes the dataset in the filesystem, then in the database.
+     *
+     * @param id The id of the dataset to delete.
+     * @param username The username of the owner of the Dataset
+     * @throws DatasetStorageException If the Dataset couldn't be deleted
+     */
+    @Override
+    public void deleteDataset(Long id, String username) throws DatasetStorageException {
+        // Delete Dataset Directories and files from the filesystem
+        try {
+            fileSystemService.deleteDatasetDirectories(id, username);
+        } catch (SecurityException e) {
+            throw new DatasetStorageException("Read access to directory was denied", e);
+        }
+        // Then delete Dataset and its files from the database
+        Optional<Dataset> dataset = datasetRepository.findById(id);
+        if (dataset.isPresent()) {
+            Set<DataFile> files = dataset.get().getFiles();
+            for (DataFile file : files) {
+                dataFileRepository.deleteById(file.getId());
+            }
+        } else {
+            throw new DatasetStorageException("Dataset removed from filesystem not found in database");
+        }
+        datasetRepository.deleteById(id);
+    }
+
+    /**
+     * Finds a dataset in the database given a particular id.
+     *
+     * @param id The id of the dataset.
+     * @return Returns the IDataset.
+     */
+    @Override
+    public IDataset findById(Long id) {
+        return datasetRepository.findById(id).orElse(null);
+    }
+
+    /**
+     * Finds all datasets in the database (Likely for admin use).
+     *
+     * @return Returns Page lists of all Datasets in the database.
+     */
+    @Override
+    public Page<Dataset> findAll() {
+        Pageable pageable = PageRequest.of(0, 20);
```

Comment:
same as above, what if user wants more than the first 20. Also, if the method is called `findAll` it should return all (or except a page request).